src/admin_ws.rs

```diff
@@ -0,0 +1,187 @@
+use std::{
+    future::Future,
+    sync::{Arc, Mutex},
+};
+
+use holochain_client::{AdminWebsocket, ConductorApiError, ConductorApiResult};
+use tokio::time::{sleep, Duration};
+
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+
+const ADMIN_WS_CONNECTION_MAX_RETRIES: usize = 1;
+const ADMIN_WS_CONNECTION_RETRY_DELAY_MS: u64 = 1000;
+
+/// A wrapper around AdminWebsocket that automatically handles reconnection
+/// when the connection is lost due to network issues or other failures.
+#[derive(Clone)]
+pub struct ReconnectingAdminWebsocket {
+    /// The WebSocket URL to connect to
+    url: String,
+    /// The handle to the AdminWebsocket connection
+    connection_handle: Arc<Mutex<Option<AdminWebsocket>>>,
+    /// Current retry attempt counter
+    current_retries: usize,
+}
+
+impl ReconnectingAdminWebsocket {
+    /// Creates a new ReconnectingAdminWebsocket with the specified parameters
+    pub fn new(url: &str) -> Self {
+        ReconnectingAdminWebsocket {
+            url: url.to_string(),
+            connection_handle: Arc::new(Mutex::new(None)),
+            current_retries: 0,
+        }
+    }
+
+    /// Checks if there is an active connection
+    fn is_connected(&self) -> HcHttpGatewayResult<bool> {
+        let connection = self.connection_handle.lock().map_err(|e| {
+            HcHttpGatewayError::InternalError(format!(
+                "Mutex was poisoned during is_connected check: {}",
+                e
+            ))
+        })?;
+
+        Ok(connection.is_some())
+    }
+
+    /// Gets reconnection retry count
+    pub fn get_reconnection_retries(&self) -> usize {
+        self.current_retries
+    }
+
+    /// Ensures that a connection is established before proceeding
+    async fn ensure_connected(&mut self) -> HcHttpGatewayResult<()> {
+        if self.is_connected()? {
+            return Ok(());
+        }
+
+        self.connect().await
```

Comment:
We had to take out a lock in `is_connected`, then another one to open the connection and store it in the mutex, and then we return which means we have to take out a third lock to use the created object. There is a potential for race conditions to do it this way, and taking out locks should be considered expensive.

This should be possible within a single lock:
1. Take out a lock
2. If there is a connection, clone and return it
3. If there is no connection, establish one and store a clone, then return the original